Adopt eslint rules in client/utils/** & migrate to typescript

lb- opened the pull request:
### General clean up of `client/utils` folder with linting re-enabled
- move all `text` utils into one file (cleanForSlug) into text.ts (general text utils)
- adopt typescript in most utils files (except performance, which is not used unless explicitly commented out for component redux work)
- add unit tests missing for various functions within utils
- add JSDOC if suitable
- fix bug with ngettext
- create noop and identity functions to use where set up in other files
- disable some rules for unit test files to make writing tests simpler, we were basically ignoring the same kinds of things across many files and this did not add value for us
- rewire the version util to remove need for extensive classes addition for throwing errors and setting up constants
- [x] Do the tests still pass?
- [X] Does the code comply with the style guide? 
- [X] **Please list the exact browser and operating system versions you tested**: Chrome 100/ Firefox 99

**Please describe additional details for testing this change**. 
- Manually retested anything that was touched even if code is basically 1:1; sidebar expanding, version notification, auto-slug generation for page title
- May be good to do a smoke test of a few of these things

gasman commented on client/src/entrypoints/admin/page-editor.js:
```diff
@@ -1,5 +1,5 @@
 import $ from 'jquery';
-import { cleanForSlug } from '../../utils/cleanForSlug';
```
Can we settle on (and document) a convention before changing this please? The current de-facto convention of "use camelcase for things that get imported into other JS, and kebab-case for endpoints" feels like it makes sense to me.

lb- replied:
ok, I was not aware of that convention, kind of makes sense when I look at the files again (ish).

I am happy to rework this PR adopting that convention to avoid having to get a decision on that yet - it still feels like a bit of a confusing situation to me.

We also use `lower_snake_case` in one scenario for entrypoints `client/src/entrypoints/contrib/typed_table_block/typed_table_block.js`

And `client/src/includes/tabs.js` even though it exports `initTabs` but `client/src/includes/initTooltips.js` exports `initTooltips` along with `initModernDropdown`.

I have flagged this question with Thibaud on the draft UI development docs.

lb- replied:
Done - kept as `cleanForSlug` and not `clean-for-slug`.

lb- commented:
Went in a bit of a different direction, instead of resolving the filename issue re-looked at the existing structure and it makes more sense to bring the slug util into the text utils folder.

This is now a much smaller functional change overall (aside from JSDoc & unit tests), less import path changes.

I have also now added unit tests to most of the other utils folder (except `performance` which is dev only) and found a bug in `ngettext` which is now resolved.

lb- commented:
updated with latest changes to `utils/version` and introduced two additional global changes to eslint

* disabled max classes rule - this does not really align with our direction of it being ok for one file to do a few things, nor does it work for cases where we actually need more classes within one file.
* refined the lines between class members - when using typescript and also static methods, it feels nicer to have these bunched instead of spaced out, so I have adjusted the rule globally accordingly to be less strict for one line class members.

gasman reviewed (changes requested):
Afraid it's pretty hard to review this with so many different things going on in the same commit - please could you create separate PRs for linting fixes, moving / renaming files, and bugfixes?

lb- commented:
Can do. Great idea.

lb- commented:
**1**

* PR https://github.com/wagtail/wagtail/pull/8746 created for the small issue in actions.ts
* branch https://github.com/lb-/wagtail/tree/cleanup/8731-eslint-fix-utils-version created for after https://github.com/wagtail/wagtail/pull/8738 is merged
* PR https://github.com/wagtail/wagtail/pull/8747 created for the gettext file (note: not really a linting issue)

lb- commented:
Closing - all other parts have been split out to already merged commits or new PRs - see updated note above https://github.com/wagtail/wagtail/pull/8578#issuecomment-1166211648

Thanks Matt - looking at this again I realise it was a bit of a random grab-bag of things that would have been really hard to review.
